src/sinks/aws_s3/partitioner.rs

```diff
@@ -0,0 +1,33 @@
+use crate::{internal_events::TemplateRenderingFailed, template::Template};
+use vector_core::{event::Event, partition::Partitioner};
+
+/// Partitions items based on the generated S3 object key for the given event.
+///
+/// TODO: Realistically, this could be a generic "template partitioner", since
+/// I'm guessing other sinks might want to partition based on a key generated
+/// from event data.
```

Comment:
I do think this could be pulled up but I'm inclined to not open an issue yet, per my reasoning [here](https://github.com/timberio/vector/pull/8884/files#r703858426).

Reply:
https://github.com/timberio/vector/issues/9078